File: blub/src/blubMain/kotlin/compile/code_analysis/ast_analizer/GetDecls.kt
package compile.code_analysis.ast_analizer

import compile.code_analysis.AstAnalyzer
import compile.code_analysis.CodeAnalyzer
import compile.parser.Stmt
import compilerError

object AstDecls {
    var functions: Array<Stmt.FunctionDecl> = arrayOf()
    var structs: Array<Stmt.StructDecl> = arrayOf()
    var structHashmap: HashMap<String, Int> = hashMapOf()
}

class GetDecls: AstAnalyzer {
    override fun analyse(codeAnalyzer: CodeAnalyzer) {
        val functions = arrayListOf<Stmt.FunctionDecl>()
        val structs = arrayListOf<Stmt.StructDecl>()

        for (stmt in codeAnalyzer.ast) {
            when (stmt) {
                is Stmt.FunctionDecl -> {
                    functions.add(stmt)
                }
                is Stmt.StructDecl -> {
                    if (AstDecls.structHashmap.contains(stmt.name))
                        compilerError("struct with name ${stmt.name} is already defined")
                    structs.add(stmt)
                    AstDecls.structHashmap[stmt.name] = structs.lastIndex
                }
                else -> {}
            }
        }
        AstDecls.functions = functions.toTypedArray()
        AstDecls.structs = structs.toTypedArray()
    }
}

File: blub/src/compile/code_analysis/ast_analyzer/get_decls.rs
use std::collections::HashMap;

use crate::{
    blub_compile_error,
    compile::{code_analysis::code_analyzer::CodeAnalyzerData, parser::ast::Stmt},
};

#[derive(Default)]
pub struct GetDecl {
    pub fns: Vec<Stmt>,
    pub fn_map: HashMap<String, usize>,
    pub structs: Vec<Stmt>,
    pub struct_map: HashMap<String, usize>,
}

impl GetDecl {
    pub fn analize(&mut self, code_analyzer: &CodeAnalyzerData) {
        for stmt in code_analyzer.ast {
            match stmt.clone() {
                Stmt::FuncDecl { name, .. } => {
                    self.fns.push(stmt.clone());
                    self.fn_map.insert(name, self.fns.len() - 1);
                }
                Stmt::StructDecl { name, .. } => {
                    if self.struct_map.contains_key(&name) {
                        blub_compile_error!("struct with name {} is already defined", &name);
                    }
                    self.structs.push(stmt.clone());
                    self.struct_map.insert(name, self.structs.len() - 1);
                }
                _ => (),
            }
        }
    }
}
